pixeltable/io/external_store.py

```diff
@@ -50,32 +69,106 @@ class Project(ExternalStore, abc.ABC):
     An `ExternalStore` that represents a labeling project. Extends `ExternalStore` with a few
     additional capabilities specific to such projects.
     """
-    def __init__(self, name: str, raw_col_mapping: Optional[dict[tuple[UUID, int], str]]):
+    def __init__(self, name: str, col_mapping: dict[Column, str], stored_proxies: Optional[dict[Column, Column]]):
         super().__init__(name)
-        self._raw_col_mapping = raw_col_mapping
-        self._col_mapping: Optional[dict[Column, str]] = None
+        self._col_mapping = col_mapping
+
+        # A mapping from original columns to proxy columns. For each entry (k, v) in the dict, `v` is the stored
+        # proxy column for `k`. The proxy column `v` must necessarily be defined in this table, but `k` need not be;
+        # `k` might instead be defined in a base table.
+        # We need to track this here, not as a property of `Column`, because the same base column might have different
+        # more than one proxy in different views (and those proxies are not interchangeable, because the views might
+        # select different rows).
+        # Note from aaron-siegel: This methodology is inefficient in the case where a table has many views with a high
+        # proportion of overlapping rows, all proxying the same base column.
+        if stored_proxies is None:
+            self.stored_proxies = {}
+        else:
+            self.stored_proxies = stored_proxies
+
+    def get_local_columns(self) -> list[Column]:
+        return list(self.col_mapping.keys())
+
+    def link(self, tbl_version: TableVersion, conn: sql.Connection) -> None:
+        # All of the media columns being linked need to either be stored computed columns, or else have stored proxies.
+        # This ensures that the media in those columns resides in the media store.
+        # First determine which columns (if any) need stored proxies, but don't have one yet.
+        stored_proxies_needed: list[Column] = []
+        for col in self.col_mapping.keys():
+            if col.col_type.is_media_type() and not (col.is_stored and col.is_computed) and col not in self.stored_proxies:
```

Comment:
How is self.stored_proxies not empty until link() gets called? (I feel I'm missing something non-obvious here...)